sdk/appconfiguration/azure-spring-cloud-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/implementation/ConnectionManager.java

```diff
@@ -0,0 +1,330 @@
+        this.currentReplica = configStore.getEndpoint();
+
+        return currentClientEndpoint;
+    }
+
+    void setCurrentClient(String replicaEndpoint) {
+        this.currentReplica = replicaEndpoint;
+    }
+
+    /**
+     * @return the originEndpoint
+     */
+    String getOriginEndpoint() {
+        return originEndpoint;
+    }
+
+    /**
+     * Returns a client.
+     * @return ConfiguraitonClient
+     */
+    List<ConfigurationClientWrapper> getAvalibleClients() {
+        if (client == null && clients == null) {
+            buildClients();
+            if (client == null && clients == null) {
+                this.health = AppConfigurationStoreHealth.NOT_LOADED;
+            }
+        }
+
+        List<ConfigurationClientWrapper> avalibleClients = new ArrayList<>();
+
+        if (client != null) {
+            avalibleClients.add(client);
+        } else if (clients != null) {
+            for (ConfigurationClientWrapper wrapper : clients) {
+                if (wrapper.getBackoffEndTime().isBefore(Instant.now())) {
+                    LOGGER.debug("Using Client: " + wrapper.getEndpoint());
+                    avalibleClients.add(wrapper);
+                }
+            }
+            if (avalibleClients.size() == 0) {
+                this.health = AppConfigurationStoreHealth.DOWN;
+            } else {
+                this.health = AppConfigurationStoreHealth.UP;
+            }
+        }
+
+        return avalibleClients;
+    }
+
+    List<String> getAllEndpoints() {
+        if (client != null) {
+            return Arrays.asList(client.getEndpoint());
+        }
+        return clients.stream().map(client -> client.getEndpoint()).collect(Collectors.toList());
+    }
+
+    /**
+     * Call when the current client failed
+     * @param endpoint replica endpoint
+     */
+    void backoffClient(String endpoint) {        
+        for (ConfigurationClientWrapper client: clients) {
+            if (client.getEndpoint().equals(endpoint)) {
+                int failedAttempt = client.getFailedAttempts();
+                long backoffTime = BackoffTimeCalculator.calculateBackoff(failedAttempt,
+                    appProperties.getDefaultMaxBackoff(),
+                    appProperties.getDefaultMinBackoff());
+                client.updateBackoffEndTime(Instant.now().plusNanos(backoffTime));
+                break;
+            }
+        }
+    }
+
+    /**
+     * Builds all the clients for a connection.
+     * @throws IllegalArgumentException when more than 1 connection method is given.
+     */
+    private void buildClients() {
+        // Single client or Multiple?
+        // If single call buildClient
+        int hasSingleConnectionString = StringUtils.hasText(configStore.getConnectionString()) ? 1 : 0;
+        int hasMultiEndpoints = configStore.getEndpoints().size() > 0 ? 1 : 0;
+        int hasMultiConnectionString = configStore.getConnectionStrings().size() > 0 ? 1 : 0;
+
```

Comment:
Could you help explain the relationship between configStore, endpoints, and connection strings?

Reply:
A ConfigStore can have many replications. Each replication has a unique endpoint. The original store replica, the origin has an endpoint format of <storename>.<domain>, each replica has an endpoint of <storename>-<replicaname>.<domain>.

Each endpoint has a unique connection string that can be used for authentication, instead of another form of Identity. If you used endpoints for connecting, then connection string can't be used.